worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java

```diff
@@ -0,0 +1,206 @@
+                    transformWorkflowResult = transformWorkflow(document, extractedProperties);
+                    if (transformWorkflowResult == null) {
+                        LOG.warn("Failure during workflow transformation. Processing of this document will not proceed " +
+                                "for this worker.");
+                        return;
+                    }
+                    // Cache the transform result
+                    workflowCache.addTransformWorkflowResult(workflowId, projectId, transformWorkflowResult);
+                }
+            }
+            finally {
+                transformWorkflowLock.unlock();
+            }
+        }
+
+        // Evaluate the document against the workflow to execute the first matching action. If document has to be sent to another
+        // worker its queue property will be updated to reflect this.
+        try
+        {
+            WorkflowEvaluator.evaluate(document, transformWorkflowResult.getTransformedWorkflow(),
+                    transformWorkflowResult.getWorkflowStorageRef());
+        } catch (final PostProcessingFailedException e) {
+            LOG.error("A failure occurred trying to evaluate document against transformed Workflow.", e);
+            document.addFailure(WorkflowWorkerConstants.ErrorCodes.WORKFLOW_EVALUATION_FAILED, e.getMessage());
+        }
+    }
+
+    /**
+     * Store the provided workflow in the data store.
+     * @param workflowJavaScript workflow to store.
+     * @param outputPartialReference partial reference to use when storing.
+     * @return storage reference for stored workflow.
+     * @throws DataStoreException if there is a failure storing workflow.
+     */
+    private String storeWorkflow(String workflowJavaScript, String outputPartialReference)
+            throws DataStoreException
+    {
+        return dataStore.store(workflowJavaScript.getBytes(StandardCharsets.UTF_8), outputPartialReference);
+    }
+
+    /**
+     * Retrieves a workflow referenced in @{code extractedProperties} and creates a script representing that workflow,
+     * storing it in the data store.
+     * @param document the current document being processed. Any failures during transformation will be recorded on the document.
+     * @param extractedProperties contains the properties required in order to transform and store the workflow.
+     * @return the result of the workflow transformation or null if a failure occurred.
+     */
+    private TransformWorkflowResult transformWorkflow(final Document document, final ExtractedProperties extractedProperties) {
+        final String workflowJavaScript;
+        try {
+            workflowJavaScript = WorkflowTransformer.retrieveAndTransformWorkflowToJavaScript(
+                    extractedProperties.getWorkflowId(), extractedProperties.getProjectId(),
+                    processingApiUrl);
+        } catch (final ApiException | WorkflowTransformerException e) {
```

Comment:
So you're saying for some `ApiException`s we should throw a `DocumentWorkerTransientException` and for others we shouldn't.  How can we differentiate between these?

Reply:
Following discussion, the plan is that when an ApiException is thrown we will call the processing API health check. If that also throws an ApiException we will throw a transient exception. If it succeeds then we will retry the retrieve and transform once, adding a failure to the document if an ApiException occurs again.